akka-javasdk/src/main/scala/akka/javasdk/impl/agent/FunctionTools.scala

```diff
@@ -39,6 +43,103 @@ object FunctionTools {
     def returnType: Class[_]
   }
 
+  /**
+   * INTERNAL API
+   */
+  @InternalApi
+  private case class RegularFunctionToolInvoker(method: Method, instanceFactory: () => Any)
+      extends FunctionToolInvoker {
+
+    private val cls = method.getDeclaringClass
+
+    override def paramNames: Array[String] =
+      method.getParameters.map(_.getName)
+
+    override def types: Array[Type] =
+      method.getGenericParameterTypes
+
+    override def invoke(args: Array[Any]): Any = {
+      try {
+        if (isAgent(cls)) method.setAccessible(true)
+        val instance = instanceFactory()
+        method.invoke(instance, args: _*)
+      } catch unwrapInvocationTargetException()
+    }
+
+    override def returnType: Class[_] =
+      method.getReturnType
+  }
+
+  private val uniqueId = "uniqueId"
+
+  /**
+   * INTERNAL API
+   */
+  @InternalApi
+  private case class EntityFunctionToolInvoker(method: Method, componentClient: ComponentClient)
+      extends FunctionToolInvoker {
+
+    private val cls = method.getDeclaringClass
+    private val hasParams: Boolean = method.getParameters.nonEmpty
+
+    override def paramNames: Array[String] = {
+      uniqueId +: method.getParameters.map(_.getName)
+    }
+
+    override def types: Array[Type] =
+      classOf[String] +: method.getGenericParameterTypes
+
+    override def invoke(args: Array[Any]): Any = {
```

Comment:
could you verify (with a sample) that tracing is also propagated all the way to the called component

Reply:
Hmm, I need to check this again. I run the tracing sample and another sample I built myself to test this PR and I was not able to see the tracing in Jaeger unless I added `akka.runtime.tracing.dev-mode = true` to the application.conf.